### PR TITLE
feat(TDP-2193 TDP-2190) : change phone number format and trim on table

### PR DESCRIPTION
**Pull request by tlnd-cgall**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-2190
https://jira.talendforge.org/browse/TDP-2193

**Please check if the PR fulfills these requirements**
- [ ] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [ ] The new code does not introduce new technical issues (sonar / eslint)
- [ ] Functional tests have been performed
- [ ] Docker configuration files for config-std or config-cloud profiles are impacted

**Please check the browsers you've tested on**
- [ ] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

### CODE REVIEW

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/52/) [:x:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/52/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/53/) [:x:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/53/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/49/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/49/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/55/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/55/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/57/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/57/)

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java [hunk 1]
```diff
@@ -227,9 +237,18 @@ public boolean acceptField(ColumnMetadata column) {
         return Type.STRING.equals(Type.get(column.getType())) || Type.INTEGER.equals(Type.get(column.getType()));
     }
 
+    @Override
+    public ActionDefinition adapt(ScopeCategory scope) {
+        return new FormatPhoneNumber(scope);
+    }
+
     @Override
     public Set<Behavior> getBehavior() {
```
Comment (aleboulanger) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java]: `getBehavior` function could be moved to `AbstractMultiScopeAction` (duplicated code in Trim and FormatPhoneNumber classes).

You can also define the behavior set during the construction of the instance:

```java
public abstract class AbstractMultiScopeAction {
    private final ScopeCategory scope;
    private final EnumSet<Behavior> behaviors;

    protected AbstractMultiScopeAction(ScopeCategory scope) {
        this.scope = scope;
        this.behaviors = EnumSet.of(ScopeCategory.DATASET.equals(scope) ? Behavior.VALUES_ALL : Behavior.VALUES_COLUMN);
    }

    @Override
    public Set<Behavior> getBehavior() {
        return behaviors;
    }
}
```

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java [hunk 2]
```diff
@@ -154,7 +162,9 @@ private String formatIfValid(String regionParam, String formatType, String phone
     @Nonnull
     public List<Parameter> getParameters(Locale locale) {
         final List<Parameter> parameters = super.getParameters(locale);
-        parameters.add(ActionsUtils.getColumnCreationParameter(locale, CREATE_NEW_COLUMN_DEFAULT));
+        if (ScopeCategory.COLUMN.equals(scope)) {
```
Comment (aleboulanger) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java]: duplicate code with Trim class. I guess you can override the getParameters method in the abstract class and put this kind of code here:

```java
public abstract class AbstractMultiScopeAction /**[...]*/ {
    @Override 
    public List<Parameter> getParameters(Locale locale) {
        List<Parameter> parameters = super.getParameters(locale);

        if (ScopeCategory.COLUMN.equals(scope)) {
            parameters.add(ActionsUtils.getColumnCreationParameter(locale, CREATE_NEW_COLUMN_DEFAULT));
        }
        
        return parameters;
    }
}
```

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/text/Trim.java [hunk 3]
```diff
@@ -17,54 +17,67 @@
...
      */
-    public static final String CUSTOM = "custom"; //$NON-NLS-1$
+    protected static final String CUSTOM = "custom"; //$NON-NLS-1$
 
-    public static final String WHITESPACE = "whitespace"; //$NON-NLS-1$
+    private static final String WHITESPACE = "whitespace"; //$NON-NLS-1$
 
     protected static final String NEW_COLUMN_SUFFIX = "_trim";
 
     private static final boolean CREATE_NEW_COLUMN_DEFAULT = false;
```
Comment (aleboulanger) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/text/Trim.java]: duplicate code with FormatPhoneNumber, move to abstract class and make it protected

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java (hunk 1 above)
Comment (tlnd-cgall) [reply dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java]: Imo, that's not the good way to do. All actions don't got the same behavior, so it's better to keep this in action's code

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java (hunk 2 above)
Comment (tlnd-cgall) [reply dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/phonenumber/FormatPhoneNumber.java]: All actions didn't got the createNewColumn, so it's not really the good way to do it imo.

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/76/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/76/)

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/text/Trim.java [hunk 4]
```diff
@@ -112,26 +128,34 @@ public void compile(ActionContext actionContext) {
...
-                row.set(ActionsUtils.getTargetColumnId(context), stringTrimmer.removeTrailingAndLeading(toTrim, parameters.get(CUSTOM_PADDING_CHAR_PARAMETER)));
-            } else {
-                row.set(ActionsUtils.getTargetColumnId(context), stringTrimmer.removeTrailingAndLeadingWhitespaces(toTrim));
-            }
+    public void apply(DataSetRow row, String columnId, String targetColumnId, ActionContext context) {
+        String toTrim = row.get(columnId);
+        row.set(targetColumnId, doTrim(toTrim, context));
+    }
+
+    private String doTrim(String toTrim, ActionContext context) {
```
Comment (aleboulanger) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/text/Trim.java]: the core behavior is in a private method which is not testable. 
Maybe move this code in the caller function

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/90/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/90/)

**Pull request merged**